### PR TITLE
feat(flagsmith-provider): Add flagsmith provider

### PR DESCRIPTION
**Pull request by gagantrivedi**
## This PR

- Adds [flagsmith](https://flagsmith.com/) provider

### Related Issues

N/A

### Notes

N/A
### Follow-up Tasks

N/A(?)

### How to test

Includes unit tests

### CODE REVIEW

**Comment by gagantrivedi** [comment]
> Hey @gagantrivedi, thanks for your contribution. Please let me know when you're ready for a review.

Hi @beeme1mr  it's ready for review now

File: providers/flagsmith/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import (
+    flagsmithClient "github.com/Flagsmith/flagsmith-go-client/v2"
+    of "github.com/open-feature/go-sdk/pkg/openfeature"
+    flagsmith "github.com/open-feature/go-sdk-contrib/providers/flagsmith/pkg"
+)
+    ...
+    // Intialise the flagsmith client
+	client := flagsmithClient.NewClient(os.Getenv("FLAGSMITH_ENVIRONMENT_KEY"))
+
+    // Inlitalise the flagsmith provider
```
Comment (beeme1mr) [inline providers/flagsmith/README.md:27]:
```suggestion
    // Initialize the flagsmith provider
```

File: providers/flagsmith/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+Here's an example of how you can use the Flagsmith provider:
+
+```go
+import (
+    flagsmithClient "github.com/Flagsmith/flagsmith-go-client/v2"
+    of "github.com/open-feature/go-sdk/pkg/openfeature"
+    flagsmith "github.com/open-feature/go-sdk-contrib/providers/flagsmith/pkg"
+)
+    ...
+    // Intialise the flagsmith client
```
Comment (beeme1mr) [inline providers/flagsmith/README.md:24]:
```suggestion
    // Initialize the flagsmith client
```

**Comment by Kavindu-Dodan** [comment]
@gagantrivedi Thank you for the contribution. I think you need to add a new entry to `.release-please-manifest.json` [1]. This controls the release please based versioning. 

[1] - https://github.com/open-feature/go-sdk-contrib/blob/main/.release-please-manifest.json

File: providers/flagsmith/go.mod [hunk 3]
```diff
@@ -0,0 +1,19 @@
+module github.com/open-feature/go-sdk-contrib/providers/flagsmith
+
+go 1.20
```
Comment (Kavindu-Dodan) [inline providers/flagsmith/go.mod:3]: Are you targeting go version 1.20? Seems your client has support for 1.18 [1], so I assume this was unintentional :) 

[1] - https://github.com/Flagsmith/flagsmith-go-client/blob/main/go.mod#L3

File: providers/flagsmith/pkg/provider.go [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+
+}
+
+// flagsmith provider does not have any hooks, returns empty slice
+func (p *Provider) Hooks() []of.Hook {
+	return []of.Hook{}
+}
+
+// Metadata returns value of Metadata (name of current service, exposed to openfeature sdk)
+func (provider *Provider) Metadata() of.Metadata {
```
Comment (Kavindu-Dodan) [inline providers/flagsmith/pkg/provider.go:36]: Let's use same receiver name 

```suggestion
func (p *Provider) Metadata() of.Metadata {
```

File: providers/flagsmith/pkg/provider.go [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+	return of.Metadata{
+		Name: "Flagsmith",
+	}
+}
+
+const TraitsKey = "traits"
```
Comment (Kavindu-Dodan) [inline providers/flagsmith/pkg/provider.go:42]: Minor - It's good to have the constant at the top of the file. ex:- [1]

[1] - https://github.com/open-feature/go-sdk-contrib/blob/main/providers/flagd/pkg/provider.go#L22

File: providers/flagsmith/pkg/provider.go [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+
+func (p *Provider) resolveFlag(ctx context.Context, flag string, defaultValue interface{}, evalCtx of.FlattenedContext) of.InterfaceResolutionDetail {
```
Comment (Kavindu-Dodan) [inline providers/flagsmith/pkg/provider.go:44]: Noticed that `ctx` is unused. It's up to you to decide to keep it or to remove it depending on your requirements

File: providers/flagsmith/pkg/provider.go [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+	resolutionDetails := res.ResolutionDetail()
+
+	if res.Error() != nil || resolutionDetails.Reason == of.DisabledReason {
+		return of.FloatResolutionDetail{
+			Value:                    defaultValue,
+			ProviderResolutionDetail: res.ProviderResolutionDetail,
+		}
+	}
+
+	misMatachResolutionErr := of.NewTypeMismatchResolutionError(fmt.Sprintf("flagsmith: Value %v is not a valid float", res.Value))
```
Comment (Kavindu-Dodan) [inline providers/flagsmith/pkg/provider.go:226]: typo - misMatchResolutionErr

**Comment by Kavindu-Dodan** [comment]
Other than the my comments, looks good in my view 👍

**Review by toddbaert — APPROVED** [review]
Looks great to me from a provider/sdk/spec perspective. I think I agree with all of @Kavindu-Dodan 's points though. Also please note [this](https://github.com/open-feature/go-sdk-contrib/pull/128#issuecomment-1460446862).

**Comment by toddbaert** [comment]
> @gagantrivedi Thank you for the contribution. I think you need to add a new entry to `.release-please-manifest.json` [1]. This controls the release please based versioning.
> 
> [1] - https://github.com/open-feature/go-sdk-contrib/blob/main/.release-please-manifest.json

It's not STRICTLY necessary, but if it's not added here it will be a `1.0`.

**Review by matthewelwell — COMMENTED** [review]
Looks good. Added one comment regarding consistency with other Flagsmith providers.

File: providers/flagsmith/go.mod (hunk 3 above)
Comment (gagantrivedi) [reply providers/flagsmith/go.mod]: Ah, yes, very good point

File: providers/flagsmith/pkg/provider.go (hunk 6 above)
Comment (gagantrivedi) [reply providers/flagsmith/pkg/provider.go:45]: Yeah, I think I would like to keep it(I have plans of updating our(flagsmith) go client to add support for this)

**Comment by matthewelwell** [comment]
> @matthewelwell, do you have any more concerns? If not, I'll merge this today and cut a release. Thanks for putting this together, @gagantrivedi!

No more concerns from me, looks good! Thanks @beeme1mr @gagantrivedi .

**Pull request merged**